src/briefcase/integrations/docker.py

```diff
@@ -378,9 +380,282 @@ def check_output(self, args: list[SubprocessArgT], image_tag: str) -> str:
+        :param cwd: The working directory for the command to be executed
+        :param add_hosts: DNS mappings to add to the container's network
+        :returns: A list of arguments that can be used to invoke the command
+            inside a docker container.
+        """
+        docker_args = ["docker", "run", "--rm"]
+
+        # Add "-it" if in interactive mode
+        if interactive:
+            docker_args.append("-it")
+
+        # Add volume mounts
+        # The :z suffix on volume mounts allows SELinux to modify the host mount; it
+        # is ignored on non-SELinux platforms
+        if mounts:
+            for source, target in mounts:
+                docker_args.extend(["--volume", f"{source}:{target}:z"])
+
+        # Pass environment variables in as --env arguments to Docker
+        if env:
+            for key, value in env.items():
+                docker_args.extend(["--env", f"{key}={value}"])
+
+        # Set a cwd as the working directory for the container
+        if cwd:
+            docker_args.extend(["--workdir", cwd])
+
+        # Host mappings to add to container's /etc/hosts
+        if add_hosts:
+            for host_name, address in add_hosts:
+                docker_args.extend(("--add-host", f"{host_name}:{address}"))
+
+        # Add the image name to create the temporary container with
+        docker_args.append(image_tag)
+
+        # Finally, add the command (and its arguments) to run in the container
+        docker_args.extend(args)
+
+        return docker_args
+
+    @contextlib.contextmanager
+    def x11_passthrough(self, subprocess_kwargs: dict[str, ...]) -> dict[str, ...]:
+        """Manager to expose the host's X11 server to a container.
+
+        This allows Docker containers to use the host's X11 server with the
+        authorization afforded to the current user. The user's X11 auth is copied and
+        modified in a dedicated temporary xauth file. A TCP server is set up to spoof a
+        new X display and proxies X commands to the current display. The XAUTHORITY
+        environment variable is set to the temporary xauth file and DISPLAY is set to
+        "host.docker.internal:<display number>" so that the container sends X commands
+        through a Docker-provided mapping of the host's network interface to the spoofed
+        display that finally proxies those commands to the actual display.
+
+        :param subprocess_kwargs: Existing keywords args from the caller
+        :returns: augmented keyword args for the call to subprocess
+        """
+        if not (DISPLAY := self.tools.os.getenv("DISPLAY")):
+            raise BriefcaseCommandError(
+                "The DISPLAY environment variable must be set to run an app in Docker"
+            )
+
+        # Create a TCP proxy for a spoofed X display
+        proxy_process, proxy_display_num = self._x11_tcp_proxy(DISPLAY)
+        xauth_file_path = self.tools.base_path / f".docker.xauth.{proxy_display_num}"
```

Comment:
IIRC, I moved this out of `/tmp/` because of Docker Desktop; it maintains a superset of directories that bind mounts can use to expose directories/files to containers. I think a recent local install of Docker Desktop for me did not include `/tmp/`....so, it just errored when Briefcase tried to start the container.

I agree, though, about using Briefcase's `tools` directory. I think Docker Desktop includes the user's `home` directory by default.....so, there might be that but I'll investigate further.

Reply:
Indeed, by default, Docker Desktop limits bind mounts to directories/files under the user's `home` directory.

> By default the `/home/<user>` directory is shared. If your project is outside this directory then it must be added to the list, otherwise you may get `Mounts denied` or `cannot start service` errors at runtime.

<sup>source: https://docs.docker.com/desktop/settings/linux/#file-sharing</sup>

So, if this temporary file is moved out of the Briefcase tools directory, we are limited to somewhere in the user's `home` directory....which probably just means the root of the user's home directory.

Otherwise, we'd need to drop an error and try to get users to add the `/tmp` directory to their Docker Desktop settings.

Reply:
I updated this to use the `build` directory in the CWD...which should be the projects `build` directory.